pkg/ext/certs.go

```diff
@@ -1,22 +1,367 @@
+	template := x509.Certificate{
+		SerialNumber: serial,
+		Subject: pkix.Name{
+			CommonName: fmt.Sprintf("%s@%d", host, time.Now().Unix()),
+		},
+		NotBefore: validFrom,
+		NotAfter:  validFrom.Add(maxAge),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	if ip := netutils.ParseIPSloppy(host); ip != nil {
+		template.IPAddresses = append(template.IPAddresses, ip)
+	} else {
+		template.DNSNames = append(template.DNSNames, host)
+	}
+
+	derBytes, err := x509.CreateCertificate(cryptorand.Reader, &template, caCertificate, &priv.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Generate cert, followed by ca
+	certBuffer := bytes.Buffer{}
+	if err := pem.Encode(&certBuffer, &pem.Block{Type: CertificateBlockType, Bytes: derBytes}); err != nil {
+		return nil, nil, err
+	}
+	if err := pem.Encode(&certBuffer, &pem.Block{Type: CertificateBlockType, Bytes: caDERBytes}); err != nil {
+		return nil, nil, err
+	}
+
+	// Generate key
+	keyBuffer := bytes.Buffer{}
+	if err := pem.Encode(&keyBuffer, &pem.Block{Type: keyutil.RSAPrivateKeyBlockType, Bytes: x509.MarshalPKCS1PrivateKey(priv)}); err != nil {
+		return nil, nil, err
+	}
+
+	return certBuffer.Bytes(), keyBuffer.Bytes(), nil
+}
+
+var _ dynamiccertificates.SNICertKeyContentProvider = &rotatingSNIProvider{}
+var _ dynamiccertificates.Notifier = &rotatingSNIProvider{}
+var _ dynamiccertificates.CertKeyContentProvider = &rotatingSNIProvider{}
+
+type rotatingSNIProvider struct {
+	name       string
+	secretName string
+
+	sninames []string
+
+	listeners []dynamiccertificates.Listener
+
+	expiresAfter time.Duration
+
+	cert []byte
+	key  []byte
+
+	secrets wranglercorev1.SecretController
+}
+
+func NewSNIProviderForCname(name string, cnames []string, secrets wranglercorev1.SecretController) (*rotatingSNIProvider, error) {
+	content := &rotatingSNIProvider{
+		name:       name,
+		secretName: fmt.Sprintf("%s-cert-ca", name),
+
+		sninames: cnames,
+
+		expiresAfter: time.Hour * 24 * 90,
+
+                secrets: secrets,
 	}
 
 	return content, nil
 }
+
+func (p *rotatingSNIProvider) AddListener(listener dynamiccertificates.Listener) {
+	p.listeners = append(p.listeners, listener)
+}
+
+func (p *rotatingSNIProvider) SNINames() []string {
+	return p.sninames
+}
+
+func (p *rotatingSNIProvider) Name() string {
+	return p.name
+}
+
+func (p *rotatingSNIProvider) CurrentCertKeyContent() ([]byte, []byte) {
+	return bytes.Clone(p.cert), bytes.Clone(p.key)
+}
+
+func (p *rotatingSNIProvider) Run(stopChan <-chan struct{}) error {
+	logrus.Info("starting imperative api cert rotator")
+
+	req, err := labels.NewRequirement(SecretLabelProvider, selection.Equals, []string{p.name})
+	if err != nil {
+		return fmt.Errorf("failed to create label requirement: %w", err)
+	}
+
+	watcher, err := p.secrets.Watch(Namespace, metav1.ListOptions{
```

Comment:
We already have a secret controller so I think we should make use of wrangler here. Something like:

```go
func Register() {
  rotatingSNIProvider := {...}
   wranglerContext.Core.Secret().OnChange(...rotatingSNIProvider.Sync)
}

func (p *rotatingSNIProvider) Sync(...) {
   if not our secret {
     return early
   }
   update listener
}
```

That would be more idiomatic to the rest of the codebase.

Reply:
Resolving this, created this issue to tackle this in the future as tech-debt: https://github.com/rancher/rancher/issues/49200.